### PR TITLE
file upload tests

### PR DESCRIPTION
**Pull request by domdorn**

### CODE REVIEW

File: src/test/scala/mockws/FileTest.scala [hunk 1]
```diff
@@ -0,0 +1,481 @@
...
+
+      val response : WSResponse = await(ws.url("/post").post(file))
+      response.body shouldEqual "POST:application/octet-stream: this is a test"
+      response.header("content-type") shouldBe Some("text/plain")
+
+      ws.close()
+    } finally {
+      file.deleteOnExit()
+    }
+  }
```
Comment (yanns) [inline src/test/scala/mockws/FileTest.scala]: You can extract the following code
```
val file = File.createTempFile("mockws", ".bin")
try {
  [...]
} finally {
  file.deleteOnExit()
}
```
into:
```
private def withTempFile[R](extension: String)(f: (File) => R): R = {
  val file = File.createTempFile("mockws", extension)
  try f(file)
  finally {
    file.deleteOnExit()
  }
}
```
and then use it:
```
  withTempFile(".bin") { file =>
    [...]
  }
```

File: src/main/scala/mockws/FakeWSRequestHolder.scala [hunk 2]
```diff
@@ -58,12 +59,19 @@ case class FakeWSRequestHolder(
...
 
-  def withMethod(method: String): Self = copy(method = method)
+  def withMethod(method: String): Self = copy(method = method.toUpperCase)
 
   def withHeaders(hdrs: (String, String)*): Self = {
     val headers = hdrs.foldLeft(this.headers)(
       (m, hdr) =>
-        if (m.contains(hdr._1)) m.updated(hdr._1, m(hdr._1) :+ hdr._2)
+        if (m.contains(hdr._1)) {
+          if(hdr._1.toLowerCase.equals("content-type")) {
```
Comment (yanns) [inline src/main/scala/mockws/FakeWSRequestHolder.scala]: try to respect the existing code formatting:
```
if(hdr._1.toLowerCase.equals("content-type")) {
```
->
```
if (hdr._1.toLowerCase.equals("content-type")) {
```

File: src/main/scala/mockws/FakeWSRequestHolder.scala [hunk 3]
```diff
@@ -188,8 +196,9 @@ case class FakeWSRequestHolder(
 
   override def options(): Future[Response] = execute("OPTIONS")
 
-  override def patch(body: File): Future[Response] =
+  override def patch(body: File): Future[Response] = {
     withBody(body).execute("PATCH")
+  }
```
Comment (yanns) [inline src/main/scala/mockws/FakeWSRequestHolder.scala]: no need for `{` `}` for simple expressions.

File: src/main/scala/mockws/FakeWSRequestHolder.scala [hunk 4]
```diff
@@ -209,7 +218,24 @@ case class FakeWSRequestHolder(
   override def withBody[T](body: T)(implicit ev: BodyWritable[T]): Self =
     withBodyAndContentType(InMemoryBody(ev.transform(body)), ev.contentType)
 
-  override def withBody(file: java.io.File): Self = copy(body = FileBody(file))
+  override def withBody(file: java.io.File): Self = {
+    if (headers.contains("Content-Type")) {
+      copy(body = FileBody(file))
+    } else {
+      var contentTypeString = URLConnection.guessContentTypeFromName(file.getName)
```
Comment (yanns) [inline src/main/scala/mockws/FakeWSRequestHolder.scala]: I am wondering if we are putting too much logic in the `MockWS` here.
Can we just simply let the user set the expected `Content-Type` header in the `Action`?

[the remark this answers is not in the document]
Comment (sahilahmadlone) [reply src/main/scala/mockws/FakeWSRequestHolder.scala]: And I will prefer some pattern matching here with an exhaustive match.

File: src/test/scala/mockws/FileTest.scala [hunk 5]
```diff
@@ -0,0 +1,481 @@
...
+
+/**
+  * Tests that [[MockWS]] simulates a WS client, in particular the methods involving authentication
+  */
+class FileTest extends FunSuite with Matchers with PropertyChecks {
+
+  val sendFileContentBackRaw = Action {
+    req =>
+      // we're expecting files here, thats why we're doing get on the optional
+      val firstFile = req.body.asRaw.get.asFile
```
[the remark this answers is not in the document]
Comment (sahilahmadlone) [reply src/test/scala/mockws/FileTest.scala]: Prefer using folds in this situation or recover.

**Comment by yanns** [comment]
no progress on this since a long time.
I close this for now.
Feel free to open it again.

**Pull request closed without merging**